### PR TITLE
Properly cache Images on the disk

### PR DESCRIPTION
**Pull request by AnnaBanani**
Properly cache Images on the disk

### CODE REVIEW

File: TheMET/TheMET/Cache/ImageCache.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        NotificationCenter.default.addObserver(self, selector: #selector(cleanUpCache), name: UIApplication.didReceiveMemoryWarningNotification, object: nil)
+    }
+    
+    @objc
+    private func cleanUpCache() {
+        self.loadedImagesCollection = [:]
+    }
+
+    private var loadedImagesCollection: [URL : UIImage] = [:]
+    private let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
```
Comment (Ssimboss) [inline TheMET/TheMET/Cache/ImageCache.swift:25]: This property is unused

File: TheMET/TheMET/Cache/ImageCache.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    
+    func loadImage(urlString: String, completion: @escaping (UIImage?) -> Void) {
+        guard let imageURL = URL(string: urlString) else {
+            completion(nil)
+            return
+        }
+        if let image = self.loadedImagesCollection[imageURL]  {
+            completion(image)
+        } else {
+            guard let fileURL = self.makeFileUrl(urlString: urlString),
```
Comment (Ssimboss) [inline TheMET/TheMET/Cache/ImageCache.swift:35]: Don’t read file-data from the disk on main thread/queue

File: TheMET/TheMET/Cache/ImageCache.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            completion(image)
+        }
+    }
+    
+    func putLoadedImageToCache(image: UIImage, for urlString: String) {
+        guard let imageURL = URL(string: urlString) else {
+            return
+        }
+        self.loadedImagesCollection[imageURL] = image
+        guard let fileURL = self.makeFileUrl(urlString: urlString) else { return }
```
Comment (Ssimboss) [inline TheMET/TheMET/Cache/ImageCache.swift:51]: Don’t write data to file on main thread/queue

**Review by Ssimboss — CHANGES_REQUESTED** [review]

File: TheMET/TheMET/Cache/ImageCache.swift (hunk 1 above)
Comment (AnnaBanani) [reply TheMET/TheMET/Cache/ImageCache.swift:25]: +

File: TheMET/TheMET/Cache/ImageCache.swift [hunk 4]
```diff
@@ -32,14 +31,16 @@ class ImageCache {
...
-                return
+            DispatchQueue.global().sync {
+                guard let fileURL = self.makeFileUrl(urlString: urlString),
+                      let data = try? Data(contentsOf: fileURL),
+                      let image = UIImage(data: data) else {
+                    completion(nil)
+                    return
+                }
+                self.loadedImagesCollection[imageURL] = image
+                completion(image)
```
Comment (Ssimboss) [inline TheMET/TheMET/Cache/ImageCache.swift:42]: The completions must be executed back on main-thread.

**Review by Ssimboss — CHANGES_REQUESTED** [review]

File: TheMET/TheMET/Cache/ImageCache.swift [hunk 5]
```diff
@@ -39,7 +39,9 @@ class ImageCache {
                     return
                 }
                 self.loadedImagesCollection[imageURL] = image
-                completion(image)
+                DispatchQueue.main.async {
+                    completion(image)
```
Comment (Ssimboss) [inline TheMET/TheMET/Cache/ImageCache.swift:43]: All the completions executed which are currently executed on `DispatchQueue.global` in this thread must be executed back on `main`

**Review by Ssimboss — CHANGES_REQUESTED** [review]

**Review by Ssimboss — APPROVED** [review]

**Pull request merged**